Correct test expectations

dougbu opened the pull request:
### Correct test expectations
- use valid `multipart/form-data` content type; include a `boundary`
- correct expectations of `FormCollection` model binder's operation
 - restore tests actually skipped for either of the above reasons
- add more tests with `ModelBindingResult.Model==null` and both `IsModelSet` values

#### Remove test references to Won't Fix bug #2473
- restore #2473 tests; update test expectations
- expect `null` composite results whenever binding fails

#### Restore test skipped due to "#2646"
- that issue does not exist; was likely #2466 or similar fixed bug

#### Rename model binding tests that still mention `ReturnsFalse` or `ReturnsTrue`

#### Minor src changes
- remove unused variable and unnecessary nesting in `DefaultControllerActionArgumentBinder`
- remove dangling mention of `[DefaultValue]` in `ComplexModelDtoModelBinder`

#### nits:
- remove empty delegates from some `GetOperationBindingContext` calls; `null` fine
- do some `using` cleanup
- combine two test methods in `KeyValuePairModelBinderTest`
- name a few more arguments

dougbu commented:
/cc @harshgMSFT @rynowak

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/BindingSourceModelBinderTest.cs:
```diff
@@ -84,12 +84,12 @@ public void BindingSourceModelBinder_ThrowsOnNonGreedySource()
             Assert.NotNull(result);
-            Assert.True(result.IsModelSet);
+            Assert.False(result.IsModelSet);
             Assert.True(binder.WasBindModelCoreCalled);
```
While you're doing cleanup here can you rename these tests to be more accurate? `BindingSourceModelBinder_ReturnsTrue_MatchingSource` `BindingSourceModelBinder_ReturnsFalse_NonMatchingSource`

They don't really return true or false

dougbu replied:
:ok:

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs:
```diff
@@ -445,6 +445,7 @@ public class CompositeModelBinderTest
                 .Returns(
                     delegate (ModelBindingContext context)
                     {
+                        // Note this ModelBindingResult is not valid.
```
It's not really invalid, the model value will just be ignored.

harshgMSFT replied:
With my latest changes, this test does not make sense. 
IMO lets replace it with `BindModel_DoesNotAddAValidationNode_IfModelBindingResultDoesNotContainAValidationNode`
PS: shorten the name if you want to.

dougbu replied:
Test makes sense since no `ModelValidationNode` with `isModelSet: false` can definitely happen. Or am I missing something?

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/ParameterBinding/ControllerActionArgumentBinderTests.cs:
```diff
@@ -361,7 +361,7 @@ public class ControllerActionArgumentBinderTests
                 .Setup(b => b.BindModelAsync(It.IsAny<ModelBindingContext>()))
                 .Returns(Task.FromResult(
-                    result: new ModelBindingResult(model: null, key: string.Empty, isModelSet: true)));
+                    result: new ModelBindingResult(model: null, key: string.Empty, isModelSet: false)));
```
This really changes the meaning of this test.

dougbu replied:
Not really much at all. Note I didn't change the assertions at all. My thought was to avoid testing a corner case in this general test. Few if any model binders return a result with `Model == null && IsModelSet`. But I can revert this one-liner change or rename the test if you prefer.

rynowak replied:
See `DefaultControllerActionArgumentBinder`@130 - there's code in place to specifically handle this scenario, and this is a test written specifically to target that code.

dougbu replied:
Then the test should confirm correct behaviour when `isModelSet` is _either_ `true` or `false`.

rynowak commented on test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs:
```diff
@@ -12,89 +12,11 @@ namespace Microsoft.AspNet.Mvc.IntegrationTests
-        [Theory(Skip = "ModelBindingResult should be non null if a model binder returns a non null resul #2473.")]
-        [InlineData(typeof(NullModelBinder), true)]
-        [InlineData(typeof(NullModelNotSetModelBinder), false)]
-        public async Task BindParameter_WithModelBinderType_NoData(Type modelBinderType, bool isModelSet)
```
Are these two tests dupes now?

dougbu replied:
Mostly irrelevant since the `CompositeModelBinder` should return `null` result in this case. I could change the assertions if we care about the scenario. Looked unimportant to me but I'm fine either way.

rynowak replied:
I think it would be preferrable to change the expectations to match the behavior then. It sounds like removing these tests would mean that we just don't have coverage for the scenario.

dougbu replied:
:ok:

harshgMSFT replied:
:+1: with @rynowak I would rather have tests that ensure the behavior.

rynowak commented on test/Microsoft.AspNet.Mvc.IntegrationTests/ModelBindingTestHelper.cs:
```diff
@@ -67,6 +69,21 @@ public static IObjectModelValidator GetObjectValidator(MvcOptions options = null
+            }
+
+            Assert.False(true, $"Unable to find { typeof(TModelBinder).FullName } in { nameof(MvcOptions) }.");
+        }
```
I'm really not a fan of this in the integration tests. Why is this needed? If these tests are failing then let them fail.

dougbu replied:
This meaning the assertion? It's a guard to ensure the setup does what the test expects.

rynowak replied:
I mean the whole thing `UpdateOptionsToEnsureNothingFollows` does. Why is it needed? 

If the goal is to verify the behavior of an individual model binder (returning null/not-null), we should have unit tests for that.

dougbu replied:
The goal is to verify the system pays attention to what the individual model binders return.

harshgMSFT commented on test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/CompositeModelBinderTest.cs:
```diff
@@ -291,7 +291,7 @@ public class CompositeModelBinderTest
             modelBinder
                 .Setup(mb => mb.BindModelAsync(It.IsAny<ModelBindingContext>()))
-                .Returns(Task.FromResult(new ModelBindingResult(null, "someName", true)));
+                .Returns(Task.FromResult(new ModelBindingResult(model: "some value", key: "someName", isModelSet: true)));
```
The test says, sets null value and sets the model state key.. why are we chaing it to a non null value?

harshgMSFT commented on test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/FormValueProviderFactoryTests.cs:
```diff
@@ -33,8 +33,8 @@ public void GetValueProvider_ReturnsNull_WhenContentTypeIsNotFormUrlEncoded()
-        [InlineData("multipart/form-data")]
-        [InlineData("multipart/form-data;charset=utf-8")]
+        [InlineData("multipart/form-data; boundary=----WebKitFormBoundarymx2fSWqWSd0OxQqq")]
+        [InlineData("multipart/form-data; boundary=----WebKitFormBoundarymx2fSWqWSd0OxQqq; charset=utf-8")]
```
:+1:

harshgMSFT commented on test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/KeyValuePairModelBinderTest.cs:
```diff
@@ -139,14 +139,13 @@ public class KeyValuePairModelBinderTest
-                    Assert.Equal("someName.key", mbc.ModelName);
-                    return Task.FromResult(new ModelBindingResult(null, string.Empty, true));
+                    Assert.Equal("someName.key", context.ModelName);
+                    return Task.FromResult(new ModelBindingResult(model: "not int", key: string.Empty, isModelSet: true));
```
line break here.

harshgMSFT commented on test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/BindingSourceModelBinderTest.cs:
```diff
@@ -105,7 +105,7 @@ public TestableBindingSourceModelBinder(BindingSource source)
             {
                 WasBindModelCoreCalled = true;
-                return Task.FromResult(new ModelBindingResult(null, bindingContext.ModelName, true));
+                return Task.FromResult(new ModelBindingResult(model: null, key: bindingContext.ModelName, isModelSet: false));
```
All the changes in this class seem to be cosmetic .. have we made sure that none of the other tests depend on a behavior like model is null and IsModelSet is true?

dougbu replied:
Not exactly cosmetic. I should have tested `isModelSet: false` with `model: null` as well as `isModelSet: true`. Instead I just switched `isModelSet` here. Will make sure both are tested.

harshgMSFT commented on test/Microsoft.AspNet.Mvc.IntegrationTests/FormFileModelBindingIntegrationTest.cs:
```diff
@@ -154,9 +154,8 @@ private class Address
-            Assert.NotNull(modelBindingResult); // Fails due to bug #2456
-            Assert.Null(modelBindingResult.Model);
+            // ModelBindingResult; expect null because binding failed.
+            Assert.Null(modelBindingResult);
```
This is wrong, the FormFileModelBinder should return non null MBR, because the type is IFormFile. That is what bug #2456 was about. Because of the quirkiness of the CompositeModelBinder, the result is null. Lets discuss this.

dougbu replied:
Will make sure the test name matches the expected behaviour.

harshgMSFT replied:
Thanks, discussed in person what is here makes sense.

dougbu commented:
Rebased and updated
- also changed PR description to match latest overall changes
- note additional fixes e.g. restoration of test skipped due to non-existent "#2646" issue and src nits

harshgMSFT commented on test/Microsoft.AspNet.Mvc.IntegrationTests/BinderTypeBasedModelBinderIntegrationTest.cs:
```diff
@@ -87,14 +110,14 @@ private class Person2
-        
-        // ModelBinderAttribute can be used without specifing the binder type. 
+
+        // ModelBinderAttribute can be used without specifying the binder type.
```
:dancers:

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/DefaultControllerActionArgumentBinder.cs:
```diff
@@ -81,24 +81,21 @@ public class DefaultControllerActionArgumentBinder : IControllerActionArgumentBi
+                    modelState,
+                    modelExplorer);
 
-                if (modelBindingResult.ValidationNode != null)
```
why is this removed? ValidationNode can be null if a model binder chooses not to use a MVN simply because it does not want to validate the model. I expected some tests to break, are they breaking?

rynowak replied:
It was moved up into the surrounding `if`

harshgMSFT replied:
my bad. Please ignore my comment.

harshgMSFT commented:
:watch:

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ModelBinding/ComplexModelDtoModelBinder.cs:
```diff
@@ -32,7 +32,7 @@ public sealed class ComplexModelDtoModelBinder : IModelBinder
                 if (modelBindingResult == null)
                 {
-                    // Could not bind. Add a result so MutableObjectModelBinder will check for [DefaultValue].
+                    // Could not bind. Let MutableObjectModelBinder know explicitly.
```
Not suggesting to add more to this PR, but the comment suggests this block is unnecessary. IMO correcting the comment is the right thing to do for now though :100:

harshgMSFT commented:
:shipit:

rynowak commented:
:shipit:

dougbu commented:
26795bd
